yt/utilities/answer_testing/testing_utilities.py

```diff
@@ -0,0 +1,557 @@
+    obj = cls(*obj_type[1])
+    return obj
+
+
+def compare_unit_attributes(ds1, ds2):
+    r"""
+    Checks to make sure that the length, mass, time, velocity, and
+    magnetic units are the same for two different dataset objects.
+    """
+    attrs = ("length_unit", "mass_unit", "time_unit", "velocity_unit", "magnetic_unit")
+    for attr in attrs:
+        u1 = getattr(ds1, attr, None)
+        u2 = getattr(ds2, attr, None)
+        assert u1 == u2
+
+
+def fake_halo_catalog(data):
+    filename = "catalog.0.h5"
+    ftypes = {field: "." for field in data}
+    extra_attrs = {"data_type": "halo_catalog", "num_halos": data["particle_mass"].size}
+    ds = {
+        "cosmological_simulation": 1,
+        "omega_lambda": 0.7,
+        "omega_matter": 0.3,
+        "hubble_constant": 0.7,
+        "current_redshift": 0,
+        "current_time": YTQuantity(1, "yr"),
+        "domain_left_edge": YTArray(np.zeros(3), "cm"),
+        "domain_right_edge": YTArray(np.ones(3), "cm"),
+    }
+    save_as_dataset(ds, filename, data, field_types=ftypes, extra_attrs=extra_attrs)
+    return filename
+
+
+def _create_plot_window_attribute_plot(ds, ptype, field, axis, pkwargs=None):
+    r"""
+    Convenience function used in plot_window_attribute_test.
+
+    Parameters
+    ----------
+    ds : Dataset
+        The Dataset object from which the plotting data is taken.
+
+    ptype : string
+        Type of plot to make (e.g., SlicePlot).
+
+    field : yt field
+        The field (e.g, density) to plot.
+
+    axis : int
+        The plot axis to plot or project along.
+
+    pkwargs : dict
+        Any keywords to be passed when creating the plot.
+    """
+    if ptype is None:
+        raise RuntimeError("Must explicitly request a plot type")
+    cls = getattr(pw, ptype, None)
+    if cls is None:
+        cls = getattr(particle_plots, ptype)
+    plot = cls(*(ds, axis, field), **pkwargs)
+    return plot
+
+
+def _create_phase_plot_attribute_plot(
+    data_source, x_field, y_field, z_field, plot_type, plot_kwargs=None
+):
+    r"""
+    Convenience function used in phase_plot_attribute_test.
+
+    Parameters
+    ----------
+    data_source : Dataset object
+        The Dataset object from which the plotting data is taken.
+
+    x_field : yt field
+        Field to plot on x-axis.
+
+    y_field : yt field
+        Field to plot on y-axis.
+
+    z_field : yt field
+        Field to plot on z-axis.
+
+    plot_type : string
+        Type of plot to make (e.g., SlicePlot).
+
+    plot_kwargs : dict
+        Any keywords to be passed when creating the plot.
+    """
+    if plot_type is None:
+        raise RuntimeError("Must explicitly request a plot type")
+    cls = getattr(profile_plotter, plot_type, None)
+    if cls is None:
+        cls = getattr(particle_plots, plot_type)
+    plot = cls(*(data_source, x_field, y_field, z_field), **plot_kwargs)
+    return plot
+
+
+def get_parameterization(fname):
+    try:
```

Comment:
Could you add a docstring to this function? If this is only used in this file, it should be probably prepended with an underscore.

Reply:
I added one. The function is used in, for example, the enzo tests, to help parametrize over every field in a dataset.